Error handling (#111)
ietf-wg-webtrans/draft-ietf-webtrans-http2: draft-ietf-webtrans-http2.md

flow control updates before receiving any response from the server.

## Session Termination and Error Handling {#errors}

An WebTransport session over HTTP/2 is terminated when either endpoint closes
the stream associated with the CONNECT request that initiated the session.

Prior to closing the stream associated with the CONNECT request, either endpoint
can send a CLOSE_WEBTRANSPORT_SESSION capsule with an application error code
and message to convey additional information about the reasons for the closure
of the session.

Session errors result in the termination of a session.  Errors can be reported
using the CLOSE_WEBTRANSPORT_SESSION capsule, which includes an error code and
an optional explanatory message.

An endpoint can terminate a session without sending a CLOSE_WEBTRANSPORT_SESSION
capsule by closing the HTTP/2 stream.

A stream that is reset terminates the session without providing an
application-level signal, though there will be an HTTP/2 error code.

This document reserves the following HTTP/2 error codes for use with reporting
WebTransport errors:

WEBTRANSPORT_ERROR (0xTBD):
: This generic error can be used for errors that do not have more specific error
  codes.

> WEBTRANSPORT_STREAM (0xTBD):

Author: ```suggestion
WEBTRANSPORT_STREAM_STATE_ERROR (0xTBD):
```

session-level flow control consumed by a stream is always known by both
endpoints.

> A WT_RESET_STREAM capsule MUST NOT be sent after a stream is closed or reset.

Reviewer: Should we add a note about intermediaries? A naive intermediary that converts QUIC RESET_STREAM to WT_RESET_STREAM is going to have a bad day, so maybe we need a "don't do that" sign

Author: Good call.  I added a general note as this relates to multiple frame/capsule types.

WebTransport, which is expected to attempt to gracefully terminate the session
as soon as possible.

## Capsule Ordering and Reliability

The use of an ordered and reliable transport means that a receiver does not need
to tolerate capsules that arrive out of order. This differs from QUIC in that a

> receiver is required to treat the arrival of out of order capsules rather than

Reviewer: Capsules are strongly ordered in QUIC, they're delivered on the request stream

```suggestion
receiver is required to treat the arrival of out of order frames rather than
```

being tolerant.

> For an intermediary that forwards from an unreliable transport (like

Reviewer: I would replace "reliable" with "strongly ordered". This makes it sound like you can't rely on QUIC to do its job

Author: ```suggestion
For an intermediary that forwards from an strongly-ordered transport (like
```
